remotes/docker/registry.go

```diff
@@ -202,11 +205,37 @@ func MatchAllHosts(string) (bool, error) {
 // MatchLocalhost is a host match function which returns true for
 // localhost.
 func MatchLocalhost(host string) (bool, error) {
-	for _, s := range []string{"localhost", "127.0.0.1", "[::1]"} {
-		if len(host) >= len(s) && host[0:len(s)] == s && (len(host) == len(s) || host[len(s)] == ':') {
-			return true, nil
+	// explicitly check ipv6 loopback without port number
+	// specified.
+	switch {
+	case host == "::1":
+		return true, nil
+	case host == "[::1]":
+		return true, nil
+	}
+	h, p, err := net.SplitHostPort(host)
+
+	// addrError helps distinguish between errors of form
+	// "no colon in address" and "too many colons" in address.
+	// The former is fine as the host string need not have a
+	// port. Latter needs to be handled.
+	addrError := &net.AddrError{
+		Err:  "missing port in address",
+		Addr: host,
+	}
+	if err != nil {
+		if err.Error() != addrError.Error() {
+			return false, err
+		} else { // host string without any port specified
+			h = host
 		}
+	} else if len(p) == 0 {
+		return false, errors.New("invalid host name form")
 	}
-	return host == "::1", nil
 
+	ips, err := net.LookupIP(h)
```

Comment:
Looking up DNS is slow, so we should statically compare with "localhost" when h is not an IP

Reply:
Also, please squash commits